torchmultimodal/models/albef.py

```diff
@@ -0,0 +1,244 @@
+            image_embeds,
+            text_embeds,
+            image_embeds_neg,
+            text_embeds_neg,
+            text_atts,
+            text_atts_neg,
+        )
+        return ALBEFOutput(
+            image_embeddings=image_embeds,
+            image_embeddings_m=image_embeds_m,
+            text_embeddings=text_embeds,
+            text_atts=text_atts,
+            vl_embeddings=vl_embeds,
+            similarity=similarity,
+        )
+
+    @torch.no_grad()
+    def _copy_params(self):
+        for model, model_m in [
+            [self.vision_encoder, self.vision_encoder_m],
+            [self.text_encoder, self.text_encoder_m],
+            [self.multimodal_encoder, self.multimodal_encoder_m],
+            [self.vision_proj, self.vision_proj_m],
+            [self.text_proj, self.text_proj_m],
+        ]:
+            for param, param_m in zip(model.parameters(), model_m.parameters()):
+                param_m.data.copy_(param.data)
+                param_m.requires_grad = False
+
+    def _unimodal_embeddings(self, image, text, text_atts):
+        image_embeds = self.vision_encoder(image)
+        text_embeds = self.text_encoder(text, attention_mask=text_atts)
+        image_feat = F.normalize(self.vision_proj(image_embeds[:, 0, :]), dim=-1)
+        text_feat = F.normalize(self.text_proj(text_embeds[:, 0, :]), dim=-1)
+        return image_embeds, text_embeds, image_feat, text_feat
+
+    @torch.no_grad()
+    def _momentum_embeddings(self, image, text, text_atts):
+        self._momentum_update()
+        image_embeds_m = self.vision_encoder_m(image)
+        text_embeds_m = self.text_encoder_m(text, attention_mask=text_atts)
+        image_feat_m = F.normalize(self.vision_proj_m(image_embeds_m[:, 0, :]), dim=-1)
+        text_feat_m = F.normalize(self.text_proj_m(text_embeds_m[:, 0, :]), dim=-1)
+        return image_embeds_m, image_feat_m, text_feat_m
+
+    @torch.no_grad()
+    def _momentum_update(self):
+        for model, model_m in [
+            [self.vision_encoder, self.vision_encoder_m],
+            [self.text_encoder, self.text_encoder_m],
+            [self.multimodal_encoder, self.multimodal_encoder_m],
+            [self.vision_proj, self.vision_proj_m],
+            [self.text_proj, self.text_proj_m],
+        ]:
+            for param, param_m in zip(model.parameters(), model_m.parameters()):
+                param_m.data = param_m.data * self.momentum + param.data * (
+                    1 - self.momentum
+                )
+
+    @torch.no_grad()
+    def _update_queue(self, image_feat_m, text_feat_m):
+        image_feats, text_feats, _ = _gather_embeddings_and_labels(
+            image_feat_m, text_feat_m
+        )
+        batch_size = image_feats.shape[0]
+        ptr = int(self.queue_ptr)
+        assert self.queue_size % batch_size == 0
+
+        self.image_queue[:, ptr : ptr + batch_size] = image_feats.T
+        self.text_queue[:, ptr : ptr + batch_size] = text_feats.T
```

Comment:
I liked the comment in the original repo that this is a dequeue and enqueue process, it makes it more clear what these lines of code are doing. perhaps you could add your own comments to explain these lines, or at the top of the function to explain what this is doing